File: modules/sql/app/model/queryCorrectors/SelectCorrector.java
package model.queryCorrectors;

import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.stream.Collectors;

import javax.inject.Singleton;

import model.SqlCorrectionException;
import model.SqlQueryResult;
import model.correctionResult.ColumnComparison;
import model.correctionResult.GroupByComparison;
import model.correctionResult.OrderByComparison;
import model.correctionResult.SqlExecutionResult;
import model.exercise.EvaluationFailed;
import model.exercise.EvaluationResult;
import model.exercise.FeedbackLevel;
import model.exercise.SqlExercise;
import model.exercise.Success;
import net.sf.jsqlparser.JSQLParserException;
import net.sf.jsqlparser.expression.Expression;
import net.sf.jsqlparser.parser.CCJSqlParserUtil;
import net.sf.jsqlparser.schema.Table;
import net.sf.jsqlparser.statement.select.Join;
import net.sf.jsqlparser.statement.select.PlainSelect;
import net.sf.jsqlparser.statement.select.Select;
import play.db.Database;

@Singleton
public class SelectCorrector extends QueryCorrector<Select, PlainSelect> {

  private static <T> List<String> listAsStrings(List<T> list) {
    if(list == null)
      return Collections.emptyList();
    return list.stream().map(t -> t.toString()).collect(Collectors.toList());
  }

  private EvaluationResult compareGroupByElements(PlainSelect plainUserQuery, PlainSelect plainSampleQuery) {
    if(plainUserQuery.getGroupByColumnReferences() == null && plainSampleQuery.getGroupByColumnReferences() == null)
      return null;

    List<String> userElements = listAsStrings(plainUserQuery.getGroupByColumnReferences());
    List<String> sampleElements = listAsStrings(plainSampleQuery.getGroupByColumnReferences());

    List<String> wrong = listDifference(userElements, sampleElements);
    List<String> missing = listDifference(sampleElements, userElements);

    return new GroupByComparison(missing, wrong);
  }

  private EvaluationResult compareOrderByElements(PlainSelect plainUserQuery, PlainSelect plainSampleQuery) {
    if(plainUserQuery.getOrderByElements() == null && plainSampleQuery.getOrderByElements() == null)
      return null;

    List<String> userElements = listAsStrings(plainUserQuery.getOrderByElements());
    List<String> sampleElements = listAsStrings(plainSampleQuery.getOrderByElements());

    List<String> wrong = listDifference(userElements, sampleElements);
    List<String> missing = listDifference(sampleElements, userElements);

    return new OrderByComparison(missing, wrong);
  }

  private List<String> getColumns(PlainSelect plainSelect) {
    return plainSelect.getSelectItems().stream().map(item -> item.toString()).collect(Collectors.toList());
  }

  @Override
  protected ColumnComparison compareColumns(PlainSelect userQuery, PlainSelect sampleQuery) {
    // FIXME: keine Beachtung der Groß-/Kleinschreibung bei Vergleich! -->
    // Verwendung core --> model.result.Matcher?
    List<String> userColumns = getColumns(userQuery).stream().map(q -> q.toUpperCase()).collect(Collectors.toList());
    List<String> sampleColumns = getColumns(sampleQuery).stream().map(q -> q.toUpperCase())
        .collect(Collectors.toList());

    List<String> wrongColumns = listDifference(userColumns, sampleColumns);
    List<String> missingColumns = listDifference(sampleColumns, userColumns);

    Success success = Success.NONE;
    if(wrongColumns.isEmpty() && missingColumns.isEmpty())
      success = Success.COMPLETE;

    return new ColumnComparison(success, missingColumns, wrongColumns);
  }

  @Override
  protected List<EvaluationResult> compareStatically(Select userQuery, Select sampleQuery,
      FeedbackLevel feedbackLevel) {
    PlainSelect plainUserQuery = (PlainSelect) userQuery.getSelectBody();
    PlainSelect plainSampleQuery = (PlainSelect) sampleQuery.getSelectBody();

    List<EvaluationResult> results = new LinkedList<>();

    EvaluationResult tableComparison = compareTables(plainUserQuery, plainSampleQuery);
    results.add(tableComparison);

    EvaluationResult columnComparison = compareColumns(plainUserQuery, plainSampleQuery);
    results.add(columnComparison);

    EvaluationResult whereComparison = compareWheres(plainUserQuery, plainSampleQuery);
    if(whereComparison != null)
      results.add(whereComparison);

    EvaluationResult orderByComparison = compareOrderByElements(plainUserQuery, plainSampleQuery);
    if(orderByComparison != null)
      results.add(orderByComparison);

    EvaluationResult groupByComparison = compareGroupByElements(plainUserQuery, plainSampleQuery);
    if(groupByComparison != null)
      results.add(groupByComparison);

    return results;
  }

  @Override
  protected EvaluationResult executeQuery(Database database, Select userStatement, Select sampleStatement,
      SqlExercise exercise, FeedbackLevel feedbackLevel) {
    SqlQueryResult userResult = null, sampleResult = null;

    try {
      Connection conn = database.getConnection();
      conn.setCatalog(exercise.scenario.shortName);

      ResultSet userResultSet = conn.createStatement().executeQuery(userStatement.toString());
      userResult = new SqlQueryResult(userResultSet);

      ResultSet sampleResultSet = conn.createStatement().executeQuery(sampleStatement.toString());
      sampleResult = new SqlQueryResult(sampleResultSet);

      conn.close();
    } catch (SQLException e) {
      return new EvaluationFailed(
          "Es gab einen Fehler beim Ausführen eines Statements:<p><pre>" + e.getMessage() + "</pre></p>");
    }

    return new SqlExecutionResult(feedbackLevel, userResult, sampleResult);

  }

  @Override
  protected List<String> getTables(PlainSelect userQuery) {
    List<String> userFromItems = new LinkedList<>();

    // Main table in Query
    if(userQuery.getFromItem() instanceof Table)
      userFromItems.add(((Table) userQuery.getFromItem()).getName());

    // All joined tables
    // FIXME: JOIN ON - Bedingung --> getOnExpression() als Bedingung?
    if(userQuery.getJoins() != null)
      for(Join join: userQuery.getJoins())
        if(join.getRightItem() instanceof Table)
          userFromItems.add(((Table) join.getRightItem()).getName());

    return userFromItems;
  }

  @Override
  protected Expression getWhere(PlainSelect query) {
    return query.getWhere();
  }
  
  @Override
  protected Select parseStatement(String statement) throws SqlCorrectionException {
    try {
      return (Select) CCJSqlParserUtil.parse(statement);
    } catch (JSQLParserException e) {
      throw new SqlCorrectionException(
          "Es gab folgenden Fehler beim Parsen ihres Statements: " + e.getCause().getMessage());
    } catch (ClassCastException e) {
      throw new SqlCorrectionException("Das Statement war vom falschen Typ! Erwartet wurde SELECT!");
    }
  }

}

File: modules/sql/app/model/querycorrectors/SelectCorrector.scala
package model.querycorrectors

import java.sql.Connection

import scala.collection.JavaConverters.asScalaBufferConverter
import scala.util.{ Failure, Success, Try }

import model.CommonUtils.cleanly
import model.CorrectionException
import model.exercise.SqlExercise
import model.matching.{ Match, MatchType, Matcher }
import model.sql.SqlQueryResult
import net.sf.jsqlparser.JSQLParserException
import net.sf.jsqlparser.expression.Expression
import net.sf.jsqlparser.parser.CCJSqlParserUtil
import net.sf.jsqlparser.schema.{ Column, Table }
import net.sf.jsqlparser.statement.select.{ OrderByElement, PlainSelect }
import play.db.Database

case class GroupByMatch(ua: Option[Expression], sa: Option[Expression], s: Int)
  extends Match[Expression](ua, sa, s) {
  override def analyze(ua: Expression, sa: Expression) = MatchType.SUCCESSFUL_MATCH
}

case class OrderByMatch(ua: Option[OrderByElement], sa: Option[OrderByElement], s: Int)
  extends Match[OrderByElement](ua, sa, s) {
  override def analyze(ua: OrderByElement, sa: OrderByElement) = MatchType.SUCCESSFUL_MATCH
}

object GROUP_BY_MATCHER extends Matcher[Expression, GroupByMatch](
  "Group By Elemente",
  List("Group By Statement"),
  _.asInstanceOf[Column].getColumnName == _.asInstanceOf[Column].getColumnName,
  new GroupByMatch(_, _, _)
)

object ORDER_BY_MATCHER extends Matcher[OrderByElement, OrderByMatch](
  "Order By Elemente",
  List("Order By Statement"),
  _.getExpression.toString == _.getExpression.toString,
  new OrderByMatch(_, _, _)
)

object SelectCorrector extends QueryCorrector("SELECT") {

  type Q = net.sf.jsqlparser.statement.select.Select

  def executeStatement(select: String, conn: Connection) =
    cleanly(conn.createStatement)(_.close)(q => new SqlQueryResult(q.executeQuery(select))) match {
      case Success(queryResult) => queryResult
      case Failure(e)           => throw new CorrectionException(select, s"Es gab einen Fehler bei der Ausfuehrung des Statements '$select'", e)
    }

  override def executeQuery(db: Database, userQ: Q, sampleQ: Q, exercise: SqlExercise): Try[SqlExecutionResult] =
    cleanly(db.getConnection)(_.close)(conn => {
      conn.setCatalog(exercise.scenario.shortName)
      new SqlExecutionResult(executeStatement(userQ.toString, conn), executeStatement(sampleQ.toString, conn))
    })

  def getColumns(select: Q) = select.getSelectBody.asInstanceOf[PlainSelect].getSelectItems.asScala.toList

  override def getColumnWrappers(query: Q): List[ColumnWrapper] =
    query.getSelectBody.asInstanceOf[PlainSelect].getSelectItems.asScala.map(ColumnWrapper.wrap(_)).toList

  override def getTableNames(select: Q) = getTables(select).map(_.getName)

  override def getTables(query: Q) = {
    val plain = query.getSelectBody.asInstanceOf[PlainSelect]

    val tables = if (plain.getFromItem.isInstanceOf[Table]) List(plain.getFromItem.asInstanceOf[Table]) else List.empty

    val joins = if (plain.getJoins == null) List.empty else
      plain.getJoins.asScala.filter(_.getRightItem.isInstanceOf[Table]).map(_.getRightItem.asInstanceOf[Table])

    (tables ++ joins).toList
  }

  override def getWhere(select: Q) = select.getSelectBody.asInstanceOf[PlainSelect].getWhere

  override def compareGroupByElements(userQ: Q, sampleQ: Q) = Some(GROUP_BY_MATCHER.doMatch(groupByElements(userQ), groupByElements(sampleQ)))

  override def compareOrderByElements(userQ: Q, sampleQ: Q) = Some(ORDER_BY_MATCHER.doMatch(orderByElements(userQ), orderByElements(sampleQ)))

  def orderByElements(userQ: Q) = {
    val javaOrderBys = userQ.getSelectBody.asInstanceOf[PlainSelect].getOrderByElements
    if (javaOrderBys == null) List.empty else javaOrderBys.asScala.toList
  }

  def groupByElements(query: Q) = {
    val javaGroupBys = query.getSelectBody.asInstanceOf[PlainSelect].getGroupByColumnReferences
    if (javaGroupBys == null) List.empty else javaGroupBys.asScala.toList
  }

  def parseStatement(statement: String) = try {
    CCJSqlParserUtil.parse(statement) match {
      case q: Q => q
      case o    => throw new CorrectionException(statement, s"Das Statement war vom falschen Typ ${o.getClass}! Erwartet wurde ein $queryType - Statement!")
    }
  } catch {
    case e: JSQLParserException => throw new CorrectionException(statement, "Es gab einen Fehler beim Parsen des Statements: " + statement, e)
  }

}
